src/utils.js

```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
+
 /**
    * Get a value
    * @param {Object} db - The hyperDB instance
    * @param {CryptoKey} enckey - The enc/dec AES key
    * @param {string} key - The key name
+   * @param {string} nonce - The nonce as hex string
    * @returns {Promise}
    */
-const get = async (db, encKey, key) => {
+const get = async (db, encKey, nonce, key) => {
   if (!db) throw new MasqError(ERRORS.NO_DB)
   if (!encKey) throw new MasqError(ERRORS.NO_ENCRYPTION_KEY)
-  const node = await db.getAsync(key)
+  if (!nonce) throw new MasqError(ERRORS.NO_NONCE)
+  const hashedKey = await hashKey(key, nonce)
+
+  const node = await db.getAsync(hashedKey)
   if (!node) return null
   const dec = await decrypt(encKey, node.value)
-  return dec
+  // dec contains an object with key and value
+  return dec.value
 }
 
 /**
    * Put a new value in the db
    * @param {Object} db - The hyperDB instance
    * @param {CryptoKey} enckey - The enc/dec AES key
+   * @param {string} nonce - The nonce as hex string
    * @param {string} key - The key name
    * @param {any} value - The value to insert
    * @returns {Promise}
    */
-const put = async (db, encKey, key, value) => {
+const put = async (db, encKey, nonce, key, value) => {
   if (!db) throw new MasqError(ERRORS.NO_DB)
   if (!encKey) throw new MasqError(ERRORS.NO_ENCRYPTION_KEY)
-  const enc = await encrypt(encKey, value)
-  return db.putAsync(key, enc)
+  if (!nonce) throw new MasqError(ERRORS.NO_NONCE)
+
+  let sanitizedKey = key
+
+  if (sanitizedKey[0] === '/') {
+    sanitizedKey = sanitizedKey.slice(1)
+  }
+
+  if (sanitizedKey[sanitizedKey.length - 1] === '/') {
+    sanitizedKey = sanitizedKey.slice(0, -1)
+  }
+
+  const withKeyName = {
+    key: sanitizedKey,
+    value: value
+  }
+  const enc = await encrypt(encKey, withKeyName)
+  const hashedKey = await hashKey(key, nonce)
+  // console.log('put', key, hashedKey)
+
+  return db.putAsync(hashedKey, enc)
 }
 
 /**
    * List all keys and values
    * @param {Object} db - The hyperDB instance
    * @param {CryptoKey} enckey - The enc/dec AES key
-   * @param {string} prefix - Prefix
+   * @param {string} nonce - The nonce as hex string
+   * @param {string} [prefix] - Prefix
    * @returns {Promise}
    */
-const list = async (db, encKey, prefix) => {
+const list = async (db, encKey, nonce, prefix) => {
   if (!db) throw new MasqError(ERRORS.NO_DB)
   if (!encKey) throw new MasqError(ERRORS.NO_ENCRYPTION_KEY)
+  if (!nonce) throw new MasqError(ERRORS.NO_NONCE)
 
-  const list = await db.listAsync(prefix)
+  const _prefix = prefix ? hashKey(prefix, nonce) : '/'
+  const list = await db.listAsync(_prefix)
   if (list.length === 1 && list[0].key === '' && list[0].value === null) {
     return {}
   }
-
-  const decList = await Promise.all(list.map(async (elt) => ({
-    key: elt.key,
-    value: await decrypt(encKey, elt.value)
-  })))
+  const decList = await Promise.all(list.map(async (elt) => {
+    const dec = await decrypt(encKey, elt.value)
+    return dec
```

Comment:
I think we can just use await Promise.all(list.map(elt => decrypt(encKey, elt.value)))